### ISSUE TITLE
bug in declare_design

### ISSUE THREAD

**Issue by graemeblair** [DeclareDesign/DeclareDesign#73]
When you include a bare function name in the `declare_design` pipeline it gives a weird error, i.e.:

```
my_population <- declare_population(N = N, noise = rnorm(N))

my_potential_outcomes <- declare_potential_outcomes(Y_Z_0 = noise, Y_Z_1 = noise + rnorm(N, mean = 2, sd = 2))

my_sampling <- declare_sampling(n = 250)

my_assignment <- declare_assignment(m = 25)

my_estimand <- declare_estimand(ATE = mean(Y_Z_1 - Y_Z_0))

my_estimator <- declare_estimator(Y ~ Z, estimand = my_estimand)

my_func <- function(data){
  return(1)
}

design <- declare_design(my_population,
                         my_potential_outcomes,
                         my_sampling,
                         my_estimand,
                         my_func,
                         dplyr::mutate(q = 5),
                         dplyr::mutate(q = 6),
                         my_assignment,
                         reveal_outcomes,
                         my_estimator)

Error in vapply(causal_order, attr, NA_character_, "type") : 
  values must be length 1,
 but FUN(X[[5]]) result is length 0
```

**Comment by nfultz** [comment]
This is a consequence of how we are currently building the pipeline to support a mix of function names and partial function calls - see eg https://github.com/DeclareDesign/DeclareDesign/blob/master/R/declare_design.R#L116

Consider:

```
> design <- declare_design(my_population,
+                          my_potential_outcomes,
+                          my_sampling,
+                          my_estimand,
+                          my_func)
Error in vapply(causal_order, attr, NA_character_, "type") : 
  values must be length 1,
 but FUN(X[[5]]) result is length 0
```

The only way a step gets wrapped (and tagged with correct metadata - eg is it dgp or not) is when it is eval'd, it throws an error. If the step expression resolves to a function, but it doesn't have the expected metadata, then finding the type fails (note that this type lookup is no longer needed/present on  my 
next-draft branch anyway, although the try/catch logic itself is still there).


And then there is 
```

> design <- declare_design(my_population,
+                          my_potential_outcomes,
+                          my_sampling,
+                          my_estimand,
+                          my_func() )
Error in vapply(causal_order, attr, NA_character_, "type") : 
  values must be length 1,
 but FUN(X[[5]]) result is length 0

```
Similarly, `myfunc()` does not throw an error either - it returns 1 without checking data - so it never goes down the wrap step codepath.

Consider this similar UDF:
```
my_func <- 
function(data){
  if(is.null(data)) stop("NULL DATA")
  return(sleep)
}
```

It throws an error - causing it to get wrapped.

This is a corner case where the UDF does not actually use the implied data argument - we might fail with a better error, but it's not clear that as written it doing something useful either.

**Comment by nfultz** [comment]
Related to #114

**Comment by jaspercooper** [comment]
This also seems related to #114 

Bare functions even with data arguments don't seem to be working atm:

``` r
my_population <- declare_population(N = 500, noise = rnorm(N))

my_potential_outcomes <- declare_potential_outcomes(
  Y_Z_0 = noise, Y_Z_1 = noise +
    rnorm(N, mean = 2, sd = 2))

my_assignment <- declare_assignment()

additional_data <- function(data) {
  data$Z2 <- data$Z
  return(data)
}
pos <- declare_potential_outcomes(handler = additional_data)

my_estimand <- declare_estimand(ATE = mean(Y_Z_1 - Y_Z_0))

my_estimator <- declare_estimator(Y ~ Z, estimand = my_estimand)

# Breaks:
design <- declare_design(my_population,
                         my_potential_outcomes,
                         my_estimand,
                         my_assignment,
                         additional_data,
                         reveal_outcomes,
                         my_estimator)
diagnosis <- diagnose_design(design)

# Works:
design <- declare_design(my_population,
                         my_potential_outcomes,
                         my_estimand,
                         my_assignment,
                         pos,
                         reveal_outcomes,
                         my_estimator)
diagnosis <- diagnose_design(design)
```

**Comment by nfultz** [comment]
Fixed in #150 - see https://github.com/DeclareDesign/DeclareDesign/blob/master/tests/testthat/test-declare-design.R#L70
